### ISSUE TITLE
`memoryview`s of Java arrays do not update their views

### ISSUE THREAD

**Issue by elevans** [scijava/scyjava#53]
We recently added some really cool fast array methods (https://github.com/scijava/scyjava/pull/52) to the code base. Playing around with them I discovered that the `memoryview()` of Java arrays do not update their views after the original Java array has been modified.

Here is a minimal example that creates a 1D `numpy.ndarray` and Java array of size 10 with some values, gets a `memoryview()` on both objects and wraps them with `np.frombuffer()` to access the arrays. Up to this point everything works nicely. However if you attempt to update the `jarr` array with some new values the `np_jarr` (which should be looking at the buffer) does not reflect the updated change.

```python
import imagej
import scyjava as sj
import numpy as np
from jpype import JInt

# initialize pyimagej
ij = imagej.init()
print(f"ImageJ version {ij.getVersion()}")

# create empty 1D NumPy array
arr = np.zeros([10], dtype=np.int32)
arr[0] = 5
arr[7] = 32

# create empty 1D Java array
jarr = JInt[10]
jarr[0] = 5
jarr[7] = 32

# print vals
print(f"init arr:  {arr}")
print(f"init jarr: {jarr}")
print("#################\n")

# get views
view_jarr = memoryview(jarr)
view_arr = memoryview(arr)

# check read-only status
if not view_jarr.readonly:
    print("jarr buffer is writable!")
if view_arr.readonly:
    print("arr buffer is readonly!")

# wrap with NumPy
np_jarr = np.frombuffer(view_jarr, dtype=np.int32)
np_arr = np.frombuffer(view_arr, dtype=np.int32)

# print vals
print(f"init np_arr:  {np_arr}")
print(f"init np_jarr: {np_jarr}")
print("#################\n")

# change original arrays
arr[3] = 55
jarr[3] = 55

# check for updated view
print(f"updated np_arr:  {np_arr}")
print(f"updated np_jarr: {np_jarr}")
print("#################")
```
Output result: `np_jarr` does not reflect the `jarr[3] = 55` change in the array.

```python
ImageJ version 2.9.0/1.53t
init arr:  [ 5  0  0  0  0  0  0 32  0  0]
init jarr: [5, 0, 0, 0, 0, 0, 0, 32, 0, 0]
#################

init np_arr:  [ 5  0  0  0  0  0  0 32  0  0]
init np_jarr: [ 5  0  0  0  0  0  0 32  0  0]
#################

updated np_arr:  [ 5  0  0 55  0  0  0 32  0  0]
updated np_jarr: [ 5  0  0  0  0  0  0 32  0  0]
#################
```

Interestingly the `np_jarr` and `view_jarr` objects report that they are looking at a `read-only` buffer which would indicate they're looking at the correct Java memory block. Attempting to write a value to `np_jarr` fails with a `read-only` value error, while the `np_arr` supports writing as we expect.

At this point I am suspecting this a JPype bug. I'm looking closer at how JPype is accessing/exposing the object's buffer protocol to see if there is something that makes sense. My best guess at this time is that JPype exposes `jarr`'s buffer protocol and simply transfers the buffer (maybe a copy?) with the [`PyBUF_WRITABLE`](https://docs.python.org/3/c-api/buffer.html#c.PyBUF_WRITABLE) flag set to `False`. This in my mind would mean the `memoryview()` is pointing to a copied Java array buffer marked `read-only`.

I'm still investigating what the `memoryview(jarr)` is actually point at.

**Comment by elevans** [comment]
After further reading it looks like write access to a Java array's underlying buffer was broken in `1.4.0`. See this [post](https://github.com/jpype-project/jpype/pull/1048#issuecomment-1110226614) for more details from Thrameos. While we already knew that we could not write to the buffer, perhaps these details will shed light on why the views are not updating.

On a related note, the docs and tests say we need to create the `memoryview` first and then do a cast `memoryview.cast()`. But this still doesn't work and the `memoryview` still do not update their views of the Java array buffer. I could also just be doing this whole thing wrong...

**Comment by Thrameos** [comment]
As far as I am aware Java array access is a one time copy.  When created it copies the memory once to an available page.  It can then be modified and when closed it can optionally write back to Java space once.  There do not appear to be calls in JNI to update the array in either direction.  Nor is there a call to share without copy. Though there is a flag that requests lazy sharing that is not guaranteed.  Thus memoryviews are intended primily to be read only to immutable data or a snapshot.  

I really dont understand the rational behind the JNI implementation.  Sharable buffers are present in Java but they are not Java arrays, so unless the Java library supports using buffers for backing whuch is rare it is hard to make use of.

**Comment by mkitti** [comment]
@Thrameos what you are referring to are the `Get<PrimitiveType>ArrayElements` routines, which may indeed make a copy and only synchronizes them upon release:
https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#Get_PrimitiveType_ArrayElements_routines

Part of the rationale for this is that Java arrays are not necessarily represented by contiguous memory buffers on the backend.

The Java 8 NIO alternative is to go thorugh `NewDirectByteBuffer`:
https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#NewDirectByteBuffer

In short, if we want a no-copy shared array, we need to use the `NewDirectByteBuffer` mechanism as is detailed in the jpype user guide:
https://jpype.readthedocs.io/en/latest/userguide.html?highlight=Buffer%20backed%20NumPy%20arrays#buffer-backed-numpy-arrays

I do not think there is a no-copy path from an existing Java 8 array to a native memory buffer to use with the buffer protocol.

The Java 19 mechanism is via the new, in preview foreign API:
https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/foreign/package-summary.html

**Comment by Thrameos** [comment]
Yes that is matches my understanding.  JPype supports both the copy method and the direct byte buffer method.   

I wonder how often the backend actually uses non-contiguous memory.   It would seem like automatically allocating large contiguous arrays as a special case rather than putting them in the regular garbage collection pool would be better if something is tool large to manage as regular memory.  After all if you are mixing native libraries you end up with two copies anyway.

Java has a lot of restrictions having to do with not driving the backend.   They could have given  additional JNI methods such as checking if an array is contiguous and given shared access, or methods to resync an existing array in one direction or another when copying rather that forcing full copies in both directions for each access.  Also direct byte buffers only are castable to byte[] and not other primitives which is highly restrictive.  The dangers of allowing reinterpret casting of primitives are small.   But these decisions all came form when Java was interpreted, so of course they seem crusty.

Java arrays are all concrete so each of the types including the Java 19 version are all going to be different implementations which are not interchangable.  I will do my best to support the Java 19 method when released.

**Comment by ctrueden** [comment]
Commits a62ec75b1b8ef6f6766587142765d8d856d9113a and f06fa5d0ce8eb691d45887f952334edcc601ec14 resolve this issue for us by switching from `memoryview` to `bytearray`, which always makes a read-write copy and is suitably efficient. The use of `memoryview` is not viable for us because calling `memoryview` a second time on the same Java array before deleting the first, even if that Java array is modified between calls, returns the same copied data from the first time, i.e. stale data, which is likely to lead to bugs in the future. Thanks everyone for the comments and/or research.
